File: source/SharpFlame/frmOptions.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using SharpFlame.Collections;
using SharpFlame.Colors;
using SharpFlame.Controls;
using SharpFlame.FileIO;
using SharpFlame.AppSettings;
using SharpFlame.Mapping;
using SharpFlame.Maths;

namespace SharpFlame
{
    public partial class frmOptions
    {
        private Font DisplayFont;

        private clsRGBA_sng MinimapCliffColour;
        private ColourControl clrMinimapCliffs;
        private clsRGBA_sng MinimapSelectedObjectColour;
        private ColourControl clrMinimapSelectedObjects;
        private PathSetControl objectDataPathSetControl = new PathSetControl("Object Data Directories");
        private PathSetControl tilesetsPathSetControl = new PathSetControl("Tilesets Directories");

        private KeyboardProfile ChangedKeyControls;

        private Option<KeyboardControl>[] lstKeyboardControls_Items = new Option<KeyboardControl>[] {};

        public frmOptions()
        {
            InitializeComponent();

            Icon = App.ProgramIcon;

            tilesetsPathSetControl.Dock = DockStyle.Fill;
            objectDataPathSetControl.Dock = DockStyle.Fill;
            TableLayoutPanel1.Controls.Add(tilesetsPathSetControl, 0, 0);
            TableLayoutPanel1.Controls.Add(objectDataPathSetControl, 0, 1);

            ChangedKeyControls = (KeyboardProfile)(KeyboardManager.KeyboardProfile.GetCopy(new KeyboardProfileCreator()));

            txtAutosaveChanges.Text = SettingsManager.Settings.AutoSaveMinChanges.ToStringInvariant();
            txtAutosaveInterval.Text = SettingsManager.Settings.AutoSaveMinInterval_s.ToStringInvariant();
            cbxAutosaveCompression.Checked = SettingsManager.Settings.AutoSaveCompress;
            cbxAutosaveEnabled.Checked = SettingsManager.Settings.AutoSaveEnabled;
            cbxAskDirectories.Checked = SettingsManager.Settings.DirectoriesPrompt;
            cbxPointerDirect.Checked = SettingsManager.Settings.DirectPointer;
            DisplayFont = SettingsManager.Settings.MakeFont();
            UpdateDisplayFontLabel();
            txtFOV.Text = SettingsManager.Settings.FOVDefault.ToStringInvariant();

            MinimapCliffColour = new clsRGBA_sng(SettingsManager.Settings.MinimapCliffColour);
            clrMinimapCliffs = new ColourControl(MinimapCliffColour);
            pnlMinimapCliffColour.Controls.Add(clrMinimapCliffs);

            MinimapSelectedObjectColour = new clsRGBA_sng(SettingsManager.Settings.MinimapSelectedObjectsColour);
            clrMinimapSelectedObjects = new ColourControl(MinimapSelectedObjectColour);
            pnlMinimapSelectedObjectColour.Controls.Add(clrMinimapSelectedObjects);

            txtMinimapSize.Text = SettingsManager.Settings.MinimapSize.ToStringInvariant();
            cbxMinimapObjectColours.Checked = SettingsManager.Settings.MinimapTeamColours;
            cbxMinimapTeamColourFeatures.Checked = SettingsManager.Settings.MinimapTeamColoursExceptFeatures;
            cbxMipmaps.Checked = SettingsManager.Settings.Mipmaps;
            cbxMipmapsHardware.Checked = SettingsManager.Settings.MipmapsHardware;
            txtUndoSteps.Text = SettingsManager.Settings.UndoLimit.ToStringInvariant();

            tilesetsPathSetControl.SetPaths(SettingsManager.Settings.TilesetDirectories);
            objectDataPathSetControl.SetPaths(SettingsManager.Settings.ObjectDataDirectories);
            tilesetsPathSetControl.SelectedNum = MathUtil.Clamp_int(Convert.ToInt32(SettingsManager.Settings.get_Value(SettingsManager.Setting_DefaultTilesetsPathNum)), -1,
                SettingsManager.Settings.TilesetDirectories.Count - 1);
            objectDataPathSetControl.SelectedNum = MathUtil.Clamp_int(Convert.ToInt32(SettingsManager.Settings.get_Value(SettingsManager.Setting_DefaultObjectDataPathNum)),
                -1, SettingsManager.Settings.ObjectDataDirectories.Count - 1);

            txtMapBPP.Text = SettingsManager.Settings.MapViewBPP.ToStringInvariant();
            txtMapDepth.Text = SettingsManager.Settings.MapViewDepth.ToStringInvariant();
            txtTexturesBPP.Text = SettingsManager.Settings.TextureViewBPP.ToStringInvariant();
            txtTexturesDepth.Text = SettingsManager.Settings.TextureViewDepth.ToStringInvariant();

            cbxPickerOrientation.Checked = SettingsManager.Settings.PickOrientation;

            UpdateKeyboardControls(-1);
        }

        public void btnSave_Click(Object sender, EventArgs e)
        {
            clsSettings NewSettings = (clsSettings)(SettingsManager.Settings.GetCopy(new SettingsCreator()));
            double dblTemp = 0;
            int intTemp = 0;

            if ( IOUtil.InvariantParse(txtAutosaveChanges.Text, ref dblTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_AutoSaveMinChanges,
                    new Change<UInt32>((uint)(MathUtil.Clamp_dbl(dblTemp, 1.0D, (Convert.ToDouble(UInt32.MaxValue)) - 1.0D))));
            }
            if ( IOUtil.InvariantParse(txtAutosaveInterval.Text, ref dblTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_AutoSaveMinInterval_s,
                    new Change<UInt32>((uint)(MathUtil.Clamp_dbl(dblTemp, 1.0D, (Convert.ToDouble(UInt32.MaxValue)) - 1.0D))));
            }
            NewSettings.set_Changes(SettingsManager.Setting_AutoSaveCompress, new Change<bool>(cbxAutosaveCompression.Checked));
            NewSettings.set_Changes(SettingsManager.Setting_AutoSaveEnabled, new Change<bool>(cbxAutosaveEnabled.Checked));
            NewSettings.set_Changes(SettingsManager.Setting_DirectoriesPrompt, new Change<bool>(cbxAskDirectories.Checked));
            NewSettings.set_Changes(SettingsManager.Setting_DirectPointer, new Change<bool>(cbxPointerDirect.Checked));
            NewSettings.set_Changes(SettingsManager.Setting_FontFamily, new Change<FontFamily>(DisplayFont.FontFamily));
            if ( IOUtil.InvariantParse(txtFOV.Text, ref dblTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_FOVDefault, new Change<double>(dblTemp));
            }
            NewSettings.set_Changes(SettingsManager.Setting_MinimapCliffColour, new Change<clsRGBA_sng>(MinimapCliffColour));
            NewSettings.set_Changes(SettingsManager.Setting_MinimapSelectedObjectsColour, new Change<clsRGBA_sng>(MinimapSelectedObjectColour));
            if ( IOUtil.InvariantParse(txtMinimapSize.Text, ref intTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_MinimapSize, new Change<int>(intTemp));
            }
            NewSettings.set_Changes(SettingsManager.Setting_MinimapTeamColours, new Change<bool>(cbxMinimapObjectColours.Checked));
            NewSettings.set_Changes(SettingsManager.Setting_MinimapTeamColoursExceptFeatures, new Change<bool>(cbxMinimapTeamColourFeatures.Checked));
            NewSettings.set_Changes(SettingsManager.Setting_Mipmaps, new Change<bool>(cbxMipmaps.Checked));
            NewSettings.set_Changes(SettingsManager.Setting_MipmapsHardware, new Change<bool>(cbxMipmapsHardware.Checked));
            if ( IOUtil.InvariantParse(txtUndoSteps.Text, ref intTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_UndoLimit, new Change<int>(intTemp));
            }
            SimpleList<string> tilesetPaths = new SimpleList<string>();
            SimpleList<string> objectsPaths = new SimpleList<string>();
            string[] controlTilesetPaths = tilesetsPathSetControl.GetPaths;
            string[] controlobjectsPaths = objectDataPathSetControl.GetPaths;
            for ( int i = 0; i <= controlTilesetPaths.GetUpperBound(0); i++ )
            {
                tilesetPaths.Add(controlTilesetPaths[i]);
            }
            for ( int i = 0; i <= controlobjectsPaths.GetUpperBound(0); i++ )
            {
                objectsPaths.Add(controlobjectsPaths[i]);
            }
            NewSettings.set_Changes(SettingsManager.Setting_TilesetDirectories, new Change<SimpleList<string>>(tilesetPaths));
            NewSettings.set_Changes(SettingsManager.Setting_ObjectDataDirectories, new Change<SimpleList<string>>(objectsPaths));
            NewSettings.set_Changes(SettingsManager.Setting_DefaultTilesetsPathNum, new Change<int>(tilesetsPathSetControl.SelectedNum));
            NewSettings.set_Changes(SettingsManager.Setting_DefaultObjectDataPathNum, new Change<int>(objectDataPathSetControl.SelectedNum));
            if ( IOUtil.InvariantParse(txtMapBPP.Text, ref intTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_MapViewBPP, new Change<int>(intTemp));
            }
            if ( IOUtil.InvariantParse(txtMapDepth.Text, ref intTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_MapViewDepth, new Change<int>(intTemp));
            }
            if ( IOUtil.InvariantParse(txtTexturesBPP.Text, ref intTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_TextureViewBPP, new Change<int>(intTemp));
            }
            if ( IOUtil.InvariantParse(txtTexturesDepth.Text, ref intTemp) )
            {
                NewSettings.set_Changes(SettingsManager.Setting_TextureViewDepth, new Change<int>(intTemp));
            }
            NewSettings.set_Changes(SettingsManager.Setting_PickOrientation, new Change<bool>(cbxPickerOrientation.Checked));

            SettingsManager.UpdateSettings(NewSettings);

            clsMap Map = Program.frmMainInstance.MainMap;
            if ( Map != null )
            {
                Map.MinimapMakeLater();
            }
            Program.frmMainInstance.View_DrawViewLater();

            KeyboardManager.KeyboardProfile = ChangedKeyControls;

            Finish(DialogResult.OK);
        }

        public void btnCancel_Click(Object sender, EventArgs e)
        {
            Finish(DialogResult.Cancel);
        }

        private bool AllowClose = false;

        public void frmOptions_FormClosing(object sender, FormClosingEventArgs e)
        {
            e.Cancel = !AllowClose;
        }

        //setting DialogResult in mono tries to close the form

        private void Finish(DialogResult result)
        {
            AllowClose = true;
            Program.frmOptionsInstance = null;
            if ( Modal )
            {
                DialogResult = result; //mono closes here
#if !Mono
                Close();
#endif
            }
            else
            {
                Close();
            }
        }

        public void btnFont_Click(Object sender, EventArgs e)
        {
            FontDialog FontDialog = new FontDialog();

            DialogResult Result = default(DialogResult);
            try //mono 267 has crashed here
            {
                FontDialog.Font = DisplayFont;
                FontDialog.FontMustExist = true;
                Result = FontDialog.ShowDialog(this);
            }
            catch
            {
                Result = DialogResult.Cancel;
            }
            if ( Result == DialogResult.OK )
            {
                DisplayFont = FontDialog.Font;
                UpdateDisplayFontLabel();
            }
        }

        public void btnAutosaveOpen_Click(Object sender, EventArgs e)
        {
            Program.frmMainInstance.Load_Autosave_Prompt();
        }

        private void UpdateDisplayFontLabel()
        {
            lblFont.Text = DisplayFont.FontFamily.Name + " " + Convert.ToString(DisplayFont.SizeInPoints) + " ";
            if ( DisplayFont.Bold )
            {
                lblFont.Text += "B";
            }
            if ( DisplayFont.Italic )
            {
                lblFont.Text += "I";
            }
        }

        private void UpdateKeyboardControl(int index)
        {
            lstKeyboardControls.Items[index] = GetKeyControlText((Option<KeyboardControl>)(KeyboardManager.OptionsKeyboardControls.Options[index]));
        }

        private void UpdateKeyboardControls(int selectedIndex)
        {
            lstKeyboardControls.Hide();
            lstKeyboardControls.Items.Clear();
            lstKeyboardControls_Items = new Option<KeyboardControl>[KeyboardManager.OptionsKeyboardControls.Options.Count];
            for ( int i = 0; i <= KeyboardManager.OptionsKeyboardControls.Options.Count - 1; i++ )
            {
                Option<KeyboardControl> item = (Option<KeyboardControl>)(KeyboardManager.OptionsKeyboardControls.Options[i]);
                string text = GetKeyControlText(item);
                lstKeyboardControls_Items[lstKeyboardControls.Items.Add(text)] = item;
            }
            lstKeyboardControls.SelectedIndex = selectedIndex;
            lstKeyboardControls.Show();
        }

        private string GetKeyControlText(Option<KeyboardControl> item)
        {
            string text = item.SaveKey + " = ";
            KeyboardControl control = (KeyboardControl)(ChangedKeyControls.get_Value(item));
            for ( int j = 0; j <= control.Keys.GetUpperBound(0); j++ )
            {
                Keys key = Keys.A;
                string keyText = Enum.GetName(typeof(Keys), key);
                if ( keyText == null )
                {
                    text += ((Int32)key).ToStringInvariant();
                }
                else
                {
                    text += keyText;
                }
                if ( j < control.Keys.GetUpperBound(0) )
                {
                    text += " + ";
                }
            }
            if ( control.UnlessKeys.GetUpperBound(0) >= 0 )
            {
                text += " unless ";
                for ( int j = 0; j <= control.UnlessKeys.GetUpperBound(0); j++ )
                {
                    Keys key = Keys.A;
                    string keyText = Enum.GetName(typeof(Keys), key);
                    if ( keyText == null )
                    {
                        text += ((Int32)key).ToStringInvariant();
                    }
                    else
                    {
                        text += keyText;
                    }
                    if ( j < control.UnlessKeys.GetUpperBound(0) )
                    {
                        text += ", ";
                    }
                }
            }
            if ( control != item.DefaultValue )
            {
                text += " (modified)";
            }

            return text;
        }

        public void btnKeyControlChange_Click(Object sender, EventArgs e)
        {
            if ( lstKeyboardControls.SelectedIndex < 0 )
            {
                return;
            }

            frmKeyboardControl capture = new frmKeyboardControl();
            if ( capture.ShowDialog() != DialogResult.OK )
            {
                return;
            }
            if ( capture.Results.Count == 0 )
            {
                return;
            }
            Option<KeyboardControl> keyOption = lstKeyboardControls_Items[lstKeyboardControls.SelectedIndex];
            KeyboardControl previous = (KeyboardControl)(ChangedKeyControls.get_Value(keyOption));

            Keys[] keys = new Keys[capture.Results.Count];
            for ( int i = 0; i <= capture.Results.Count - 1; i++ )
            {
                keys[i] = capture.Results[i].Item;
            }
            KeyboardControl copy = new KeyboardControl(keys, previous.UnlessKeys);
            ChangedKeyControls.set_Changes(keyOption, new Change<KeyboardControl>(copy));
            UpdateKeyboardControl(keyOption.GroupLink.ArrayPosition);
        }

        public void btnKeyControlChangeUnless_Click(Object sender, EventArgs e)
        {
            if ( lstKeyboardControls.SelectedIndex < 0 )
            {
                return;
            }

            frmKeyboardControl capture = new frmKeyboardControl();
            if ( capture.ShowDialog() != DialogResult.OK )
            {
                return;
            }
            if ( capture.Results.Count == 0 )
            {
                return;
            }
            Option<KeyboardControl> keyOption = lstKeyboardControls_Items[lstKeyboardControls.SelectedIndex];
            KeyboardControl previous = (KeyboardControl)(ChangedKeyControls.get_Value(keyOption));

            Keys[] unlessKeys = new Keys[capture.Results.Count];
            for ( int i = 0; i <= capture.Results.Count - 1; i++ )
            {
                unlessKeys[i] = capture.Results[i].Item;
            }
            KeyboardControl copy = new KeyboardControl(previous.Keys, unlessKeys);
            ChangedKeyControls.set_Changes(keyOption, new Change<KeyboardControl>(copy));
            UpdateKeyboardControl(keyOption.GroupLink.ArrayPosition);
        }

        public void btnKeyControlChangeDefault_Click(Object sender, EventArgs e)
        {
            if ( lstKeyboardControls.SelectedIndex < 0 )
            {
                return;
            }

            Option<KeyboardControl> keyOption = lstKeyboardControls_Items[lstKeyboardControls.SelectedIndex];
            ChangedKeyControls.set_Changes(keyOption, null);
            UpdateKeyboardControl(keyOption.GroupLink.ArrayPosition);
        }
    }
}

File: reference/FlaMEVB/frmOptions.vb

Public Class frmOptions

    Private DisplayFont As Font

    Private MinimapCliffColour As clsRGBA_sng
    Private clrMinimapCliffs As ctrlColour
    Private MinimapSelectedObjectColour As clsRGBA_sng
    Private clrMinimapSelectedObjects As ctrlColour
    Private ObjectDataPathSet As New ctrlPathSet("Object Data Directories")
    Private TilesetsPathSet As New ctrlPathSet("Tilesets Directories")

    Private ChangedKeyControls As clsKeyboardProfile

    Private lstKeyboardControls_Items(-1) As clsOption(Of clsKeyboardControl)

    Public Sub New()
        InitializeComponent()

        Icon = ProgramIcon

#If Mono Then
        For Each tab As TabPage In TabControl1.TabPages
            tab.Text &= " "c
        Next
#End If

        TilesetsPathSet.Dock = DockStyle.Fill
        ObjectDataPathSet.Dock = DockStyle.Fill
        TableLayoutPanel1.Controls.Add(TilesetsPathSet, 0, 0)
        TableLayoutPanel1.Controls.Add(ObjectDataPathSet, 0, 1)

        ChangedKeyControls = CType(KeyboardProfile.GetCopy(New clsKeyboardProfileCreator), clsKeyboardProfile)

        txtAutosaveChanges.Text = InvariantToString_uint(Settings.AutoSaveMinChanges)
        txtAutosaveInterval.Text = InvariantToString_uint(Settings.AutoSaveMinInterval_s)
        cbxAutosaveCompression.Checked = Settings.AutoSaveCompress
        cbxAutosaveEnabled.Checked = Settings.AutoSaveEnabled
        cbxAskDirectories.Checked = Settings.DirectoriesPrompt
        cbxPointerDirect.Checked = Settings.DirectPointer
        DisplayFont = Settings.MakeFont
        UpdateDisplayFontLabel()
        txtFOV.Text = InvariantToString_dbl(Settings.FOVDefault)

        MinimapCliffColour = New clsRGBA_sng(Settings.MinimapCliffColour)
        clrMinimapCliffs = New ctrlColour(MinimapCliffColour)
        pnlMinimapCliffColour.Controls.Add(clrMinimapCliffs)

        MinimapSelectedObjectColour = New clsRGBA_sng(Settings.MinimapSelectedObjectsColour)
        clrMinimapSelectedObjects = New ctrlColour(MinimapSelectedObjectColour)
        pnlMinimapSelectedObjectColour.Controls.Add(clrMinimapSelectedObjects)

        txtMinimapSize.Text = InvariantToString_int(Settings.MinimapSize)
        cbxMinimapObjectColours.Checked = Settings.MinimapTeamColours
        cbxMinimapTeamColourFeatures.Checked = Settings.MinimapTeamColoursExceptFeatures
        cbxMipmaps.Checked = Settings.Mipmaps
        cbxMipmapsHardware.Checked = Settings.MipmapsHardware
        txtUndoSteps.Text = InvariantToString_uint(Settings.UndoLimit)

        TilesetsPathSet.SetPaths(Settings.TilesetDirectories)
        ObjectDataPathSet.SetPaths(Settings.ObjectDataDirectories)
        TilesetsPathSet.SelectedNum = Clamp_int(CType(Settings.Value(Setting_DefaultTilesetsPathNum), Integer), -1, Settings.TilesetDirectories.Count - 1)
        ObjectDataPathSet.SelectedNum = Clamp_int(CType(Settings.Value(Setting_DefaultObjectDataPathNum), Integer), -1, Settings.ObjectDataDirectories.Count - 1)

        txtMapBPP.Text = InvariantToString_int(Settings.MapViewBPP)
        txtMapDepth.Text = InvariantToString_int(Settings.MapViewDepth)
        txtTexturesBPP.Text = InvariantToString_int(Settings.TextureViewBPP)
        txtTexturesDepth.Text = InvariantToString_int(Settings.TextureViewDepth)

        cbxPickerOrientation.Checked = Settings.PickOrientation

        UpdateKeyboardControls(-1)
    End Sub

    Private Sub btnSave_Click(sender As System.Object, e As System.EventArgs) Handles btnSave.Click

        Dim NewSettings As clsSettings = CType(Settings.GetCopy(New clsSettingsCreator), clsSettings)
        Dim dblTemp As Double
        Dim intTemp As Integer

        If InvariantParse_dbl(txtAutosaveChanges.Text, dblTemp) Then
            NewSettings.Changes(Setting_AutoSaveMinChanges) = New clsOptionProfile.clsChange(Of UInteger)(CUInt(Clamp_dbl(dblTemp, 1.0#, CDbl(UInteger.MaxValue) - 1.0#)))
        End If
        If InvariantParse_dbl(txtAutosaveInterval.Text, dblTemp) Then
            NewSettings.Changes(Setting_AutoSaveMinInterval_s) = New clsOptionProfile.clsChange(Of UInteger)(CUInt(Clamp_dbl(dblTemp, 1.0#, CDbl(UInteger.MaxValue) - 1.0#)))
        End If
        NewSettings.Changes(Setting_AutoSaveCompress) = New clsOptionProfile.clsChange(Of Boolean)(cbxAutosaveCompression.Checked)
        NewSettings.Changes(Setting_AutoSaveEnabled) = New clsOptionProfile.clsChange(Of Boolean)(cbxAutosaveEnabled.Checked)
        NewSettings.Changes(Setting_DirectoriesPrompt) = New clsOptionProfile.clsChange(Of Boolean)(cbxAskDirectories.Checked)
        NewSettings.Changes(Setting_DirectPointer) = New clsOptionProfile.clsChange(Of Boolean)(cbxPointerDirect.Checked)
        NewSettings.Changes(Setting_FontFamily) = New clsOptionProfile.clsChange(Of FontFamily)(DisplayFont.FontFamily)
        If InvariantParse_dbl(txtFOV.Text, dblTemp) Then
            NewSettings.Changes(Setting_FOVDefault) = New clsOptionProfile.clsChange(Of Double)(dblTemp)
        End If
        NewSettings.Changes(Setting_MinimapCliffColour) = New clsOptionProfile.clsChange(Of clsRGBA_sng)(MinimapCliffColour)
        NewSettings.Changes(Setting_MinimapSelectedObjectsColour) = New clsOptionProfile.clsChange(Of clsRGBA_sng)(MinimapSelectedObjectColour)
        If InvariantParse_int(txtMinimapSize.Text, intTemp) Then
            NewSettings.Changes(Setting_MinimapSize) = New clsOptionProfile.clsChange(Of Integer)(intTemp)
        End If
        NewSettings.Changes(Setting_MinimapTeamColours) = New clsOptionProfile.clsChange(Of Boolean)(cbxMinimapObjectColours.Checked)
        NewSettings.Changes(Setting_MinimapTeamColoursExceptFeatures) = New clsOptionProfile.clsChange(Of Boolean)(cbxMinimapTeamColourFeatures.Checked)
        NewSettings.Changes(Setting_Mipmaps) = New clsOptionProfile.clsChange(Of Boolean)(cbxMipmaps.Checked)
        NewSettings.Changes(Setting_MipmapsHardware) = New clsOptionProfile.clsChange(Of Boolean)(cbxMipmapsHardware.Checked)
        If InvariantParse_int(txtUndoSteps.Text, intTemp) Then
            NewSettings.Changes(Setting_UndoLimit) = New clsOptionProfile.clsChange(Of Integer)(intTemp)
        End If
        Dim tilesetPaths As New SimpleList(Of String)
        Dim objectsPaths As New SimpleList(Of String)
        Dim controlTilesetPaths() As String = TilesetsPathSet.GetPaths
        Dim controlobjectsPaths() As String = ObjectDataPathSet.GetPaths
        For i As Integer = 0 To controlTilesetPaths.GetUpperBound(0)
            tilesetPaths.Add(controlTilesetPaths(i))
        Next
        For i As Integer = 0 To controlobjectsPaths.GetUpperBound(0)
            objectsPaths.Add(controlobjectsPaths(i))
        Next
        NewSettings.Changes(Setting_TilesetDirectories) = New clsOptionProfile.clsChange(Of SimpleList(Of String))(tilesetPaths)
        NewSettings.Changes(Setting_ObjectDataDirectories) = New clsOptionProfile.clsChange(Of SimpleList(Of String))(objectsPaths)
        NewSettings.Changes(Setting_DefaultTilesetsPathNum) = New clsOptionProfile.clsChange(Of Integer)(TilesetsPathSet.SelectedNum)
        NewSettings.Changes(Setting_DefaultObjectDataPathNum) = New clsOptionProfile.clsChange(Of Integer)(ObjectDataPathSet.SelectedNum)
        If InvariantParse_int(txtMapBPP.Text, intTemp) Then
            NewSettings.Changes(Setting_MapViewBPP) = New clsOptionProfile.clsChange(Of Integer)(intTemp)
        End If
        If InvariantParse_int(txtMapDepth.Text, intTemp) Then
            NewSettings.Changes(Setting_MapViewDepth) = New clsOptionProfile.clsChange(Of Integer)(intTemp)
        End If
        If InvariantParse_int(txtTexturesBPP.Text, intTemp) Then
            NewSettings.Changes(Setting_TextureViewBPP) = New clsOptionProfile.clsChange(Of Integer)(intTemp)
        End If
        If InvariantParse_int(txtTexturesDepth.Text, intTemp) Then
            NewSettings.Changes(Setting_TextureViewDepth) = New clsOptionProfile.clsChange(Of Integer)(intTemp)
        End If
        NewSettings.Changes(Setting_PickOrientation) = New clsOptionProfile.clsChange(Of Boolean)(cbxPickerOrientation.Checked)

        UpdateSettings(NewSettings)

        Dim Map As clsMap = frmMainInstance.MainMap
        If Map IsNot Nothing Then
            Map.MinimapMakeLater()
        End If
        frmMainInstance.View_DrawViewLater()

        KeyboardProfile = ChangedKeyControls

        Finish(Windows.Forms.DialogResult.OK)
    End Sub

    Private Sub btnCancel_Click(sender As System.Object, e As System.EventArgs) Handles btnCancel.Click

        Finish(Windows.Forms.DialogResult.Cancel)
    End Sub

    Private AllowClose As Boolean = False

    Private Sub frmOptions_FormClosing(sender As Object, e As System.Windows.Forms.FormClosingEventArgs) Handles Me.FormClosing

        e.Cancel = Not AllowClose
    End Sub

    'setting DialogResult in mono tries to close the form

    Private Sub Finish(result As Windows.Forms.DialogResult)

        AllowClose = True
        frmOptionsInstance = Nothing
        If Modal Then
            DialogResult = result 'mono closes here
#If Not Mono Then
            Close()
#End If
        Else
            Close()
        End If
    End Sub

    Private Sub btnFont_Click(sender As System.Object, e As System.EventArgs) Handles btnFont.Click
        Dim FontDialog As New Windows.Forms.FontDialog

        Dim Result As DialogResult
        Try 'mono 267 has crashed here
            FontDialog.Font = DisplayFont
            FontDialog.FontMustExist = True
            Result = FontDialog.ShowDialog(Me)
        Catch
            Result = Windows.Forms.DialogResult.Cancel
        End Try
        If Result = Windows.Forms.DialogResult.OK Then
            DisplayFont = FontDialog.Font
            UpdateDisplayFontLabel()
        End If
    End Sub

    Private Sub btnAutosaveOpen_Click(sender As System.Object, e As System.EventArgs) Handles btnAutosaveOpen.Click

        frmMainInstance.Load_Autosave_Prompt()
    End Sub

    Private Sub UpdateDisplayFontLabel()

        lblFont.Text = DisplayFont.FontFamily.Name & " " & DisplayFont.SizeInPoints & " "
        If DisplayFont.Bold Then
            lblFont.Text &= "B"
        End If
        If DisplayFont.Italic Then
            lblFont.Text &= "I"
        End If
    End Sub

    Private Sub UpdateKeyboardControl(index As Integer)

        lstKeyboardControls.Items(index) = GetKeyControlText(CType(Options_KeyboardControls.Options(index), clsOption(Of clsKeyboardControl)))
    End Sub

    Private Sub UpdateKeyboardControls(selectedIndex As Integer)

        lstKeyboardControls.Hide()
        lstKeyboardControls.Items.Clear()
        ReDim lstKeyboardControls_Items(Options_KeyboardControls.Options.Count - 1)
        For i As Integer = 0 To Options_KeyboardControls.Options.Count - 1
            Dim item As clsOption(Of clsKeyboardControl) = CType(Options_KeyboardControls.Options(i), clsOption(Of clsKeyboardControl))
            Dim text As String = GetKeyControlText(item)
            lstKeyboardControls_Items(lstKeyboardControls.Items.Add(text)) = item
        Next
        lstKeyboardControls.SelectedIndex = selectedIndex
        lstKeyboardControls.Show()
    End Sub

    Private Function GetKeyControlText(item As clsOption(Of clsKeyboardControl)) As String

        Dim text As String = item.SaveKey & " = "
        Dim control As clsKeyboardControl = CType(ChangedKeyControls.Value(item), clsKeyboardControl)
        For j As Integer = 0 To control.Keys.GetUpperBound(0)
            Dim key As Keys = control.Keys(j)
            Dim keyText As String = [Enum].GetName(GetType(Keys), key)
            If keyText Is Nothing Then
                text &= InvariantToString_int(key)
            Else
                text &= keyText
            End If
            If j < control.Keys.GetUpperBound(0) Then
                text &= " + "
            End If
        Next
        If control.UnlessKeys.GetUpperBound(0) >= 0 Then
            text &= " unless "
            For j As Integer = 0 To control.UnlessKeys.GetUpperBound(0)
                Dim key As Keys = control.UnlessKeys(j)
                Dim keyText As String = [Enum].GetName(GetType(Keys), key)
                If keyText Is Nothing Then
                    text &= InvariantToString_int(key)
                Else
                    text &= keyText
                End If
                If j < control.UnlessKeys.GetUpperBound(0) Then
                    text &= ", "
                End If
            Next
        End If
        If control IsNot item.DefaultValue Then
            text &= " (modified)"
        End If

        Return text
    End Function

    Private Sub btnKeyControlChange_Click(sender As System.Object, e As System.EventArgs) Handles btnKeyControlChange.Click

        If lstKeyboardControls.SelectedIndex < 0 Then
            Exit Sub
        End If

        Dim capture As New frmKeyboardControl
        If capture.ShowDialog() <> Windows.Forms.DialogResult.OK Then
            Exit Sub
        End If
        If capture.Results.Count = 0 Then
            Exit Sub
        End If
        Dim keyOption As clsOption(Of clsKeyboardControl) = lstKeyboardControls_Items(lstKeyboardControls.SelectedIndex)
        Dim previous As clsKeyboardControl = CType(ChangedKeyControls.Value(keyOption), clsKeyboardControl)
        Dim keys() As Keys
        ReDim keys(capture.Results.Count - 1)
        For i As Integer = 0 To capture.Results.Count - 1
            keys(i) = capture.Results(i).Item
        Next
        Dim copy As clsKeyboardControl = New clsKeyboardControl(keys, previous.UnlessKeys)
        ChangedKeyControls.Changes(keyOption) = New clsOptionProfile.clsChange(Of clsKeyboardControl)(copy)
        UpdateKeyboardControl(keyOption.GroupLink.ArrayPosition)
    End Sub

    Private Sub btnKeyControlChangeUnless_Click(sender As System.Object, e As System.EventArgs) Handles btnKeyControlChangeUnless.Click

        If lstKeyboardControls.SelectedIndex < 0 Then
            Exit Sub
        End If

        Dim capture As New frmKeyboardControl
        If capture.ShowDialog() <> Windows.Forms.DialogResult.OK Then
            Exit Sub
        End If
        If capture.Results.Count = 0 Then
            Exit Sub
        End If
        Dim keyOption As clsOption(Of clsKeyboardControl) = lstKeyboardControls_Items(lstKeyboardControls.SelectedIndex)
        Dim previous As clsKeyboardControl = CType(ChangedKeyControls.Value(keyOption), clsKeyboardControl)
        Dim unlessKeys() As Keys
        ReDim unlessKeys(capture.Results.Count - 1)
        For i As Integer = 0 To capture.Results.Count - 1
            unlessKeys(i) = capture.Results(i).Item
        Next
        Dim copy As clsKeyboardControl = New clsKeyboardControl(previous.Keys, unlessKeys)
        ChangedKeyControls.Changes(keyOption) = New clsOptionProfile.clsChange(Of clsKeyboardControl)(copy)
        UpdateKeyboardControl(keyOption.GroupLink.ArrayPosition)
    End Sub

    Private Sub btnKeyControlChangeDefault_Click(sender As System.Object, e As System.EventArgs) Handles btnKeyControlChangeDefault.Click

        If lstKeyboardControls.SelectedIndex < 0 Then
            Exit Sub
        End If

        Dim keyOption As clsOption(Of clsKeyboardControl) = lstKeyboardControls_Items(lstKeyboardControls.SelectedIndex)
        ChangedKeyControls.Changes(keyOption) = Nothing
        UpdateKeyboardControl(keyOption.GroupLink.ArrayPosition)
    End Sub
End Class
